compiler/luci-interpreter/src/kernels/Add.test.cpp

```diff
@@ -27,59 +27,118 @@ namespace
 
 using namespace testing;
 
-TEST(AddTest, Float)
+// for quantized Add, the error shouldn't exceed step
+float GetTolerance(float min, float max)
 {
-  Shape base_shape = {2, 3, 1, 2};
-  std::vector<Shape> test_shapes{{1, 1, 3, 2}, {1, 3, 1, 2}, {2, 1, 3, 1}, {2, 3, 1, 1}};
-  std::vector<std::vector<float>> test_outputs = {
-      {0.0f, 2.6f, 0.0f, 2.8f, 0.7f, 3.2f, 1.1f, 0.8f, 0.5f, 1.0f, 1.9f, 1.4f,
-       1.0f, 0.0f, 0.4f, 0.0f, 1.8f, 0.0f, 1.4f, 3.1f, 0.8f, 3.3f, 2.2f, 3.7f,
-       0.0f, 0.3f, 0.0f, 0.5f, 0.0f, 0.9f, 0.9f, 0.0f, 0.3f, 0.0f, 1.7f, 0.0f},
-      {0.0f, 2.6f, 0.5f, 1.0f, 1.8f, 0.0f, 1.4f, 3.1f, 0.0f, 0.5f, 1.7f, 0.0f},
-      {0.0f, 2.5f, 0.0f, 2.6f, 0.0f, 1.9f, 1.1f, 0.7f, 1.2f, 0.8f, 0.5f, 0.1f,
-       1.0f, 0.0f, 1.1f, 0.0f, 0.4f, 0.0f, 1.7f, 3.3f, 2.2f, 3.8f, 2.1f, 3.7f,
-       0.0f, 0.5f, 0.0f, 1.0f, 0.0f, 0.9f, 1.2f, 0.0f, 1.7f, 0.0f, 1.6f, 0.0f},
-      {0.0f, 2.5f, 1.2f, 0.8f, 0.4f, 0.0f, 1.7f, 3.3f, 0.0f, 1.0f, 1.6f, 0.0f}};
-  std::vector<float> input1_data{-0.3f, 2.3f, 0.9f,  0.5f, 0.8f, -1.1f,
-                                 1.2f,  2.8f, -1.6f, 0.0f, 0.7f, -2.2f};
-  std::vector<float> input2_data{0.2f, 0.3f, -0.4f, 0.5f, 1.0f, 0.9f};
-  for (size_t i = 0; i < test_shapes.size(); ++i)
-  {
-    Tensor input1_tensor = makeInputTensor<DataType::FLOAT32>(base_shape, input1_data);
-    Tensor input2_tensor = makeInputTensor<DataType::FLOAT32>(test_shapes[i], input2_data);
-    Tensor output_tensor = makeOutputTensor(DataType::FLOAT32);
+  float kQuantizedStep = (max - min) / 255.0;
+  return kQuantizedStep;
+}
 
-    AddParams params{};
-    params.activation = Activation::RELU;
+float getMinValue(std::initializer_list<float> data)
+{
+  return 0 > std::min<float>(data) ? std::min<float>(data) : 0;
+}
 
-    Add kernel(&input1_tensor, &input2_tensor, &output_tensor, params);
-    kernel.configure();
-    kernel.execute();
+float getMaxValue(std::initializer_list<float> data)
+{
+  return 0 < std::max<float>(data) ? std::max<float>(data) : 0;
+}
 
-    EXPECT_THAT(extractTensorData<float>(output_tensor),
-                ::testing::ElementsAreArray(ArrayFloatNear(test_outputs[i], 0.0001f)))
-        << "With shape number " << i;
+template <typename T>
+void Check(std::initializer_list<int32_t> input1_shape, std::initializer_list<int32_t> input2_shape,
+           std::initializer_list<int32_t> output_shape, std::initializer_list<float> input1_data,
+           std::initializer_list<float> input2_data, std::initializer_list<float> output_data)
+{
+  float kQuantizedTolerance = GetTolerance(getMinValue(input1_data) + getMinValue(input2_data),
```

Comment:
I don't understand the meaning of this code. Isn't this element-wise add?
input 1 = [1, 2, 3]
input 2 = [3, 2, 1]
input 1 + input 2 = [4, 4, 4] (min = 4, max = 4)
According to your code, min = 2, max = 6.

Reply:
@seanshpark They are values, not shapes :)  [1, 2, 3] is a vector of length 3, whose values are 1, 2, and 3.

Reply:
in case, if input1 and input2 have a same shape, it will be [4,4,4]. but in case if input1 shape [1,3] and input2 shape [3,1], then output would be [[2,3,4],[3,4,5],[4,5,6]]. I just thought min value is a possible minimal value that inferred from input values. so, it is sum of min value in each inputs. and max value is also the same.

Reply:
Sorry for make confused. I will use the value in TFlite example value.